Extensive TOPAS workflow update:

HomolkaN opened the pull request:
# New pull request, because changes made to the dev_MC master branch had not been implemented.
## 2nd August 2022: Merged pull requests [#470](https://github.com/e0404/matRad/pull/470) and [#524](https://github.com/e0404/matRad/pull/524), Added support for 4D
#### Major changes:
1. Restructured the MCEmittanceBaseData class, fit and calculation pipeline
2. Added example 13 for generating analytical data file by fitting to given machine emittance
2.1. added function to plot particleBaseDataEntry
2.2. edited function to fit base data
2.3. added function to generate a single pencil beam
3. Added support for 4D calculation

#### Minor changes:
1. added variable in pln to contain an existing BDL file if available
2. Throws info where base data file is saved
3. renamed MCsquare property "Num_Primaries" -> "numHistories" to be in line with other Monte Carlo (this is written to the BDL file in original format)
4. Added more comments


## 15th June 2022: Added merge commit from my main branch
#### Major changes to Monte Carlo pipeline:
-> This change was done to make usage more streamlined and uniform as well as be in line with future changes
1. Merged support functions for TOPAS into topasConfig class
2. Merged support functions for MCsquare into MCsquareConfig class
3. Added a class constructor for pln in MatRad_config, which loads requested classes in pln and writes default values that were not

#### Futher changes:
1. Added comments to whole TOPAS pipeline
2. calcParticleDoseMCtopas now generates dij in matRad format
3. Changed matRad_calcCubes to accept a variety of differnet fields for Monte Carlo, without changing the current usage


## 5th December 2021
#### Major updates:
1. installTOPAS script which installs TOPAS with the dependencies from downloaded topas.tar.gz
2. modular Schneider Converter: Converter is generated on demand rather than read from file which allows a variety of different options
3. modular TOPAS scorer which can be individually turned on and off
4. added 4D accumulation in calcDoseDirectMC
5. implemented dij calculation in TOPAS

#### Minor updates:
1. Added optional initial weights to fluence Optimization
2. enabled helium in calcdoseDirect
3. implemented std of physicalDose
4. export feature for TOPAS to run externally (includes functions to read from external folders)
5. restructured resampling in calcParticleDoseMCtopas in separate function
6. MCemmittanceBaseData can calculate meanEnergy and spread also for carbon and helium ions

#### Bug fixes and minor changes:
1. added flag in stf to catch specific error where no energies could be found
2. Fixed bug in resizeCstToGrid function
3. Fixed a number of errors in 4D calculation workflow

wahln reviewed (changes requested):
Good work. Some documentation missing and some additions are not quite clear to me yet.

wahln commented:
By the way - tests are failing mainly due to an issue with the matRad_multScen implementation. Will check this within the next days...

wahln commented on matRad_calcParticleDoseMCtopas.m:
```diff
@@ -158,58 +183,130 @@
+                                dij.abx(dij.bx>0) = dij.ax(dij.bx>0)./dij.bx(dij.bx>0);
+                                
+                                dij.mAlphaDose{ctScen,1}(:,d)      = dij.physicalDose{ctScen,1}(:,d) .* dij.alpha{ctScen,1}(:,d);
+                                dij.mSqrtBetaDose{ctScen,1}(:,d)   = sqrt(dij.physicalDose{ctScen,1}(:,d)) .* dij.beta{ctScen,1}(:,d);
```
As discussed, this does not make sense - must be sqrt(beta) * physicalDose

wahln commented on topas/matRad_TopasReadExternal.m:
```diff
@@ -0,0 +1,88 @@
+                end
+                
+                dij.mAlphaDose{ctScen,1}(:,d)           = dij.physicalDose{ctScen,1}(:,d) .* dij.alpha{ctScen,1}(:,d);
+                dij.mSqrtBetaDose{ctScen,1}(:,d)        = sqrt(dij.physicalDose{ctScen,1}(:,d)) .* dij.beta{ctScen,1}(:,d);
```
again: must be sqrt(beta) * dose

HomolkaN commented:
## Added merge commit from my main branch:

#### Major changes to Monte Carlo pipeline:
-> This change was done to make usage more streamlined and uniform as well as be in line with future changes
1. Merged support functions for TOPAS into topasConfig class
2. Merged support functions for MCsquare into MCsquareConfig class
3. Added a class constructor for pln in MatRad_config, which loads requested classes in pln and writes default values that were not 

#### Futher changes:

4. Added comments to whole TOPAS pipeline
5. calcParticleDoseMCtopas now generates dij in matRad format 
6. Changed matRad_calcCubes to accept a variety of differnet fields for Monte Carlo, without changing the current usage

wahln commented:
There still seem to be some issues in testing. Maybe run the tests locally on your side @HomolkaN to identify the issues.

wahln commented on matRad_calcDoseDirectMC.m:
```diff
@@ -92,7 +92,12 @@
             matRad_cfg.dispWarning('Number of beams stored not the same as size of dij. Using singular weight for MC');
-            dij.numOfBeams = size(dij.physicalDose{1},2);
+            % This parameter should be 1 since calcDoseDirect is true and all weights are used in the simulation already
+            if size(dij.physicalDose{1},2) ~= 1
+                matRad_cfg.dispError('Singular weight was called but dij has %i columns!',1)
+            end
+            dij.numOfBeams = 1;
+            dij.beamNum = 1;
```
Hi @HomolkaN 
I don't think this override should be done here alone. The parameters need to be set in the respective functions to create a valid dij with one line. I will prepare a fix for this quickly. You then need to test if it works with TOPAS as well.

HomolkaN replied:
both `matRad_calcParticleDoseMCsquare `and `matRad_calcParticleDoseMCtopas` generate valid dijs in both calcDoseDirect and dij mode. So this override is not needed anymore.

HomolkaN commented on matRad_resizeCstToGrid.m:
```diff
@@ -35,14 +35,15 @@
 for i = 1:size(cst,1)            % loop over all structures
    for j = 1:numel(cst{i,4})     % loop over all scenarios
-      tmpCube              = zeros([numel(vXgridOld) numel(vYgridOld) numel(vZgridOld)]);
+      tmpCube              = zeros([numel(vYgridOld) numel(vXgridOld) numel(vZgridOld)]);
```
If the change is not correct, there has to be aa mistake somewhere else, because the outdated version did throw an error here.

wahln replied:
Yeah the question was just to double check, to me it makes sense: The grid's are in coordinates (x,y,z) and the cube in in indices and thus permuted: j,i,k, meaning i corresponds to y and j to x.

HomolkaN commented on matRad_calcDoseDirectMC.m:
```diff
@@ -116,11 +125,18 @@
 
+if pln.multScen.totNumScen ~= 1
+    resultGUI.accPhysicalDose = zeros(size(resultGUI.phaseDose{1}));
+    for i = 1:pln.multScen.totNumScen
```
Yes I think it makes more sense to loop over ctScen

wahln replied:
Yes. I also think that we might need to double check at a later point what's happening in ct scenarios + uncertainty (range/shift) scenarios. I am quite sure that @JenHardt will come across this during her work and could investigate.

HomolkaN commented on matRad_calcDoseDirectMC.m:
```diff
@@ -1,4 +1,4 @@
-function resultGUI = matRad_calcDoseDirectMC(ct,stf,pln,cst,w,nHistories)
+function resultGUI = matRad_calcDoseDirectMC(ct,stf,pln,cst,w,nHistories,exportForExternalCalculation)
```
The parameter has been moved to the TOPASconfig and documentation has been added

wahln replied:
By the way - I would make this a pln parameter. It's better for keeping interface compatibility.

HomolkaN replied:
With the config updates, it will show up in pln.propMC.externalCalculation. But the reason it is part of the topasConfig is that it's only needed for TOPAS, so in my view, it makes sense to be in the specifig config.

HomolkaN commented on matRad_calcParticleDoseMCtopas.m:
```diff
@@ -158,58 +183,130 @@
-                            delete(finishedFiles{runIx});
-                            topasCall = [topasCall '; touch ' finishedFiles{runIx} ' &'];
+                              
+                if exportForExternalCalculation
```
Yes, in order to read the calculated files back into matRad, meta information about e.g. the dij is needed. That information is now selectively stored in the MCparam struct.
I don't know exactly what you mean with providing a function handle, since the workflows to run topas externally are so vastly different?

wahln replied:
let's say you specify an interface for starting the simulation and gathering the files. you can say that the starting function would always needs to be a function that gets the parameters (topasFiles, folder, blablabla). Then someone could write a function getting this parameters to launch the simulation somewhere, or call some other script etc., and just provide the handle of this function: `fHandle = @myStartupFunction` and then our code calls `fHandle(topasFiles, folder, blablabla)`.

You don't need to implement this right now, but you could write a TODO comment. It would be a nice feature.

HomolkaN commented on matRad_calcCubes.m:
```diff
@@ -107,8 +121,14 @@
-            resultGUI.(tally) = reshape(full(dij.(tally){scenNum}),dij.doseGrid.dimensions);
+        if ~isfield(resultGUI,tally) && ~contains(tally,'std')
+            tallyCut = strsplit(tally,'_');
+            if size(tallyCut,2) > 1
```
This is now outdated since `matRad_calcParticleDoseMCtopas` and `matRad_calcParticleDoseMCsquare` handle this internally and all fields are processed in the new calcCubes function anyways, so there should not be any unprocessed fields in the first place.

HomolkaN commented on MatRad_MCemittanceBaseData.m:
```diff
@@ -58,6 +58,7 @@
             end
             
             matRad_cfg = MatRad_Config.instance();
+            obj.energyspread = matRad_cfg.propMC.defaultCarbonEnergySpread;
```
For now this has been rudimentary implemented in commit [4a3e73a9cb4f825214478ebfc5ef43e60ca9cc42](https://github.com/e0404/matRad/pull/536/commits/4a3e73a9cb4f825214478ebfc5ef43e60ca9cc42)
